pkg/gpu/stream.go

```diff
@@ -0,0 +1,178 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package gpu
+
+import (
+	"math"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/gpu/model"
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	gpuebpf "github.com/DataDog/datadog-agent/pkg/gpu/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// StreamHandler is responsible for receiving events from a single CUDA stream and generating
+// stats from them.
+type StreamHandler struct {
+	kernelLaunches []gpuebpf.CudaKernelLaunch
+	memAllocEvents map[uint64]gpuebpf.CudaMemEvent
+	kernelSpans    []*model.KernelSpan
+	allocations    []*model.MemoryAllocation
+	processEnded   bool // A marker to indicate that the process has ended, and this handler should be flushed
+}
+
+func newStreamHandler() *StreamHandler {
+	return &StreamHandler{
+		memAllocEvents: make(map[uint64]gpuebpf.CudaMemEvent),
+	}
+}
+
+func (sh *StreamHandler) handleKernelLaunch(event *gpuebpf.CudaKernelLaunch) {
+	// Copy events, as the memory can be overwritten in the ring buffer after the function returns
```

Comment:
Not sure I understand what this comment means.
It is true that the memory is over-written in the ring buffer, but the data you get is a copy allocated by the loader library. So this extra copy is not needed here.

Reply:
At first it was not a copy, but then I noticed I started getting corrupted events and I assumed it came from there. The issue is that I call `Done()` on the event coming from the library to return it to the pool, because the lifetime of that event being copied here is unknown.

Reply:
Is it possible to call `Done` when the lifetime of the object is over? Maybe having `CudaKernelLaunch` implement its own `Done` to return the object to the pool.

If we are performing an allocation for each record then there is no advantage of using the `sync.Pool`, since the pressure on the GC would just come from these new allocations.

Reply:
It is, but it complicates the code a bit and we might end up with events with quite a long lifetime.

The pool comes from the ring buffer event handler, is it optional? Because as you say we don't really need it.

Reply:
The reason we use a pool is just to have a cache of memory which the GC does not have to handle.
You will have to be sure that the we do not incur any extra overhead  by removing the pool. 

If we can measure that we do not, then you can opt out of using the pool by not setting the [RecordGetter](https://github.com/DataDog/datadog-agent/pull/29646/files#diff-29de7ac45aee65827bf04f46ba6bec8a60792654518642abcadbbc7c8ca1a0dcR200). That way the library will make the allocation for you.

Reply:
I think for now we can leave it like this and run some profiling once we deploy this on some GPU clusters. My worry about the long lifetime is not just about the effect on the pool but also about the possibility of leaks.

Reply:
I do not follow your reasons for not using sync.Pool here.
1. What leaks are you talking about?
2. As far as code complexity goes, is it not enough to call a `Done` function in `getCurrentKernelSpan`?

Reply:
Oh no, I meant to leave the pool usage and everything as is. In regards to both questions, my issue is with kernel launches that don't end up being included in a kernel span, so that event isn't returned to the pool. In theory it shouldn't happen, but I prefer to be sure and pay the copy price to have the ring payload lifetime well defined. Also, I'm just noticing that it helps with tests as I don't need to have different `Done()` behaviours on the event depending on whether it's a sample test event or a real one.

Reply:
> pay the copy price to have the ring payload lifetime well defined
I do not see any benefit in having the ring payload lifetime well defined, if an allocation can leak anyway. If I understand correctly, a kernel launch not included in a span would result in the copied event to remain in the list forever. The memory overhead in this case would be the same as the allocation from the pool leaking.

So you are paying the extra overhead of an allocation in the average case (kernel launch included in the span), without any benefit in the worst case (kernel launch event never gets freed).

Btw, if it is possible for such a leak to happen then we should have some way to detect them, otherwise we risk the memory usage of system-probe increasing without bounds.

Reply:
>I think for now we can leave it like this and run some profiling once we deploy this on some GPU clusters.

By the way, we do not need to wait to deploy to a GPU clusters to get this data. You can generate some synthetic workload with a mocked app and use [pprof](https://github.com/google/pprof) to see what the overhead of this extra copy is.

Reply:
I don't have a good idea either... Maybe @brycekahle knows?

Reply:
Resolving this, based on offline conversation.